File: internal/domain/progression.go
package domain

import (
	"strings"
	"time"
)

// ProgressionStrategy determines how target weight changes between sessions.
type ProgressionStrategy string

const (
	ProgressionLinear     ProgressionStrategy = "linear"
	ProgressionPercentage ProgressionStrategy = "percentage"
	ProgressionWave       ProgressionStrategy = "wave"
)

// IsValid returns true if s is a recognized progression strategy.
func (s ProgressionStrategy) IsValid() bool {
	switch s {
	case ProgressionLinear, ProgressionPercentage, ProgressionWave:
		return true
	}
	return false
}

// ValidProgressionStrategies returns the allowed strategy string values.
func ValidProgressionStrategies() []string {
	return []string{
		string(ProgressionLinear),
		string(ProgressionPercentage),
		string(ProgressionWave),
	}
}

// ProgressionStrategyMessage returns the validation error message for strategy.
func ProgressionStrategyMessage() string {
	return "must be one of: " + strings.Join(ValidProgressionStrategies(), ", ")
}

// ProgressionRule defines how target weight changes between sessions for a section exercise.
type ProgressionRule struct {
	ID                int64
	UUID              string
	SectionExerciseID int64
	Strategy          ProgressionStrategy
	Increment         *float64
	IncrementPct      *float64
	DeloadThreshold   int
	DeloadPct         float64
	CreatedAt         time.Time
	UpdatedAt         time.Time
}

// Validate enforces domain business rules on a ProgressionRule.
func (r *ProgressionRule) Validate() error {
	if !r.Strategy.IsValid() {
		return NewValidationError("strategy", ProgressionStrategyMessage())
	}
	if r.Strategy == ProgressionLinear && (r.Increment == nil || *r.Increment <= 0) {
		return NewValidationError("increment", "is required and must be positive for linear strategy")
	}
	if r.Strategy == ProgressionPercentage && (r.IncrementPct == nil || *r.IncrementPct <= 0) {
		return NewValidationError("increment_pct", "is required and must be positive for percentage strategy")
	}
	if r.DeloadThreshold < 1 {
		return NewValidationError("deload_threshold", "must be at least 1")
	}
	if r.DeloadPct < 0 || r.DeloadPct > 100 {
		return NewValidationError("deload_pct", "must be between 0 and 100")
	}
	return nil
}

// NextWeight calculates the target weight for the next session based on the
// current weight, the number of consecutive failures, and the progression rule.
func (r *ProgressionRule) NextWeight(current float64, consecutiveFailures int) float64 {
	if consecutiveFailures >= r.DeloadThreshold {
		return current * (1 - r.DeloadPct/100)
	}
	if consecutiveFailures > 0 {
		return current // hold weight on failure below deload threshold
	}
	switch r.Strategy {
	case ProgressionLinear:
		return current + *r.Increment
	case ProgressionPercentage:
		return current * (1 + *r.IncrementPct/100)
	case ProgressionWave:
		// Wave loading deferred — seeded 5/3/1 programs use static target weights.
		return current
	}
	return current
}

File: app/domain/progression.ts
// Progression domain model — mirrors internal/domain/progression.go.

import { ValidationError } from "./errors";

// ---------------------------------------------------------------------------
// Value objects
// ---------------------------------------------------------------------------

export type ProgressionStrategy = "linear" | "percentage" | "wave";

export const VALID_STRATEGIES: ProgressionStrategy[] = [
  "linear",
  "percentage",
  "wave",
];

// ---------------------------------------------------------------------------
// Entity
// ---------------------------------------------------------------------------

export interface ProgressionRule {
  id: number;
  uuid: string;
  section_exercise_id: number;
  strategy: ProgressionStrategy;
  increment: number | null;
  increment_pct: number | null;
  deload_threshold: number;
  deload_pct: number;
  created_at: string;
  updated_at: string;
}

// ---------------------------------------------------------------------------
// Validation
// ---------------------------------------------------------------------------

export function validateProgressionRule(r: {
  strategy: string;
  increment?: number | null;
  increment_pct?: number | null;
  deload_threshold: number;
  deload_pct: number;
}): ValidationError | null {
  if (!VALID_STRATEGIES.includes(r.strategy as ProgressionStrategy)) {
    return new ValidationError(
      "strategy",
      `invalid strategy: ${r.strategy}`
    );
  }

  if (r.strategy === "linear") {
    if (r.increment == null || r.increment <= 0) {
      return new ValidationError(
        "increment",
        "increment must be > 0 for linear strategy"
      );
    }
  }

  if (r.strategy === "percentage") {
    if (r.increment_pct == null || r.increment_pct <= 0) {
      return new ValidationError(
        "increment_pct",
        "increment_pct must be > 0 for percentage strategy"
      );
    }
  }

  if (r.deload_threshold < 1) {
    return new ValidationError(
      "deload_threshold",
      "deload_threshold must be >= 1"
    );
  }

  if (r.deload_pct < 0 || r.deload_pct > 100) {
    return new ValidationError(
      "deload_pct",
      "deload_pct must be between 0 and 100"
    );
  }

  return null;
}

// ---------------------------------------------------------------------------
// Core progression logic — direct port of ProgressionRule.NextWeight()
// ---------------------------------------------------------------------------

export function nextWeight(
  rule: {
    strategy: ProgressionStrategy;
    increment: number | null;
    increment_pct: number | null;
    deload_threshold: number;
    deload_pct: number;
  },
  currentWeight: number,
  consecutiveFailures: number
): number {
  // Deload: consecutive failures >= threshold
  if (consecutiveFailures >= rule.deload_threshold) {
    return currentWeight * (1 - rule.deload_pct / 100);
  }

  // Hold: some failures, but below threshold
  if (consecutiveFailures > 0) {
    return currentWeight;
  }

  // Increment: zero failures
  switch (rule.strategy) {
    case "linear":
      return currentWeight + (rule.increment ?? 0);
    case "percentage":
      return currentWeight * (1 + (rule.increment_pct ?? 0) / 100);
    case "wave":
      // Wave loading deferred — static weights only.
      return currentWeight;
  }
}
